File: src/io/github/hulang1024/chinesechess/rule/AbstractChess.java
package io.github.hulang1024.chinesechess.rule;

/**
 * 抽象的棋子
 * @author Hu Lang
 */
public abstract class AbstractChess {
    /** 游戏 */
    public ChineseChessGame game;
    /** 棋盘 */
    public Chessboard chessboard;
    /** 当前位置 */
    public ChessPosition pos;
    /** 所属军方 */
    public HostEnum host;

    /**
     * 判断指定位置是否可走。
     * 不考虑指定位置是否为空或有棋子，不应包含全局规则。
     * @param destPos
     * @return
     */
    public abstract boolean canGoTo(ChessPosition destPos);
}

File: web-client/chinese-chess/src/rule/chess/AbstractChess.ts
import Chess from "../Chess";
import ChessPos from "../ChessPos";
import ChessHost from "../chess_host";
import Game from "../Game";

/**
 * 抽象的棋子
 */
export default abstract class AbstractChess implements Chess {
    /** 当前位置 */
    protected pos: ChessPos;
    /** 所属棋方 */
    protected host: ChessHost;

    constructor(pos: ChessPos, host: ChessHost) {
        this.pos = pos;
        this.host = host;
    }

    canGoTo(destPos: ChessPos, game: Game) {
        return false;
    }

    setPos(pos: ChessPos) {
        this.pos = pos;
    }

    getPos() {
        return this.pos;
    }

    getHost() {
        return this.host;
    }

    is(chessClass: Function) {
        return this instanceof chessClass;
    }
}
